spatial/curve/hilbert_test.go

```diff
@@ -0,0 +1,381 @@
+			0x19, 0x18, 0x17, 0x14,
+
+			0x03, 0x04, 0x09, 0x0A,
+			0x02, 0x05, 0x0E, 0x0D,
+			0x1D, 0x1C, 0x11, 0x12,
+			0x1E, 0x1F, 0x16, 0x15,
+
+			0x3C, 0x3B, 0x36, 0x35,
+			0x3D, 0x3A, 0x31, 0x32,
+			0x22, 0x23, 0x2E, 0x2D,
+			0x21, 0x20, 0x29, 0x2A,
+
+			0x3F, 0x38, 0x37, 0x34,
+			0x3E, 0x39, 0x30, 0x33,
+			0x25, 0x24, 0x2F, 0x2C,
+			0x26, 0x27, 0x28, 0x2B,
+		},
+	}
+
+	for order, expected := range cases {
+		t.Run(fmt.Sprintf("Case/%d", order), func(t *testing.T) {
+			testCurveCase(t, Hilbert3D{Order: order}, order, expected)
+		})
+	}
+}
+
+func TestHilbert4D(t *testing.T) {
+	for dim := 2; dim <= 4; dim++ {
+		t.Run(fmt.Sprintf("Dimension/%d", dim), func(t *testing.T) {
+			for ord := 1; ord <= 4; ord++ {
+				t.Run(fmt.Sprintf("Order %d", ord), func(t *testing.T) {
+					testCurve(t, Hilbert4D{Order: ord})
+				})
+			}
+		})
+	}
+}
+
+func BenchmarkHilbert(b *testing.B) {
+	for N := 2; N <= 4; N++ {
+		b.Run(fmt.Sprintf("%dD/Curve", N), func(b *testing.B) {
+			for ord := 1; ord <= 10; ord++ {
+				b.Run(fmt.Sprintf("Order %d", ord), func(b *testing.B) {
+					h := newCurve(ord, N)
+					v := make([]int, N)
+					for i := range v {
+						v[i] = rand.Intn(1 << ord)
+					}
+					u := make([]int, N)
+					for n := 0; n < b.N; n++ {
+						copy(u, v)
+						h.Curve(u)
+					}
+				})
+			}
+		})
+
+		b.Run(fmt.Sprintf("%dD/Space", N), func(b *testing.B) {
+			for ord := 1; ord <= 10; ord++ {
+				b.Run(fmt.Sprintf("Order %d", ord), func(b *testing.B) {
+					h := newCurve(ord, N)
+					d := rand.Intn(1 << (ord * N))
+					for n := 0; n < b.N; n++ {
+						h.Space(d)
+					}
+				})
+			}
+		})
+	}
+}
+
+type curve interface {
+	Dims() []int
+	Len() int
+	Curve(v []int) int
+	Space(d int) []int
+}
+
+func newCurve(order, dim int) curve {
+	switch dim {
+	case 2:
+		return Hilbert2D{Order: order}
+	case 3:
+		return Hilbert3D{Order: order}
+	case 4:
+		return Hilbert4D{Order: order}
+	}
+	panic("invalid dimension")
+}
+
+func testCurve(t *testing.T, c curve) {
+	t.Helper()
+
+	var errc int
+	fail := func() {
+		if errc < 10 {
+			errc++
+			t.Fail()
+		} else {
+			t.FailNow()
+		}
+	}
+
+	m := map[int][]int{}
+	curveRange(c, func(v []int) {
+		d := c.Curve(dup(v))
+		u := c.Space(d)
+		if !reflect.DeepEqual(v, u) {
+			t.Logf("Space is not the inverse of Curve for d=%d %v", d, v)
+			fail()
+		}
+
+		m[d] = dup(v)
+	})
+
+	D := 1
+	for _, v := range c.Dims() {
+		D *= v
+	}
+	for d := 0; d < D-1; d++ {
+		v, u := m[d], m[d+1]
+		if !adjacent(v, u) {
+			t.Logf("points %x and %x are not adjacent", d, d+1)
+			t.Logf("    %v -> %v", v, u)
+			fail()
+		}
+	}
+}
+
+func curveRange(c curve, fn func([]int)) {
+	size := c.Dims()
+	dimRange(len(size), size, make([]int, len(size)), fn)
+}
+
+func dimRange(dim int, size []int, v []int, fn func([]int)) {
+	if dim == 0 {
+		fn(v)
+		return
+	}
+
+	for i := 0; i < size[dim-1]; i++ {
+		v[dim-1] = i
+		dimRange(dim-1, size, v, fn)
+	}
+}
+
+// v may get mangled by Curve, so copy it.
```

Comment:
This documents how/where/why the function is used, not what it does. But probably it doesn't need a comment at all.